File: drafting_service/evaluator.py
"""Evaluation layer — synced 1-to-1 with draft_poc/evaluator.py.

Single-Stage Programmatic Evaluation:
  Rule-based checks (fast, transparent, fail-fast):
  - Length (words for email, characters for LinkedIn connection note cap)
  - Readability (Flesch Reading Ease & Flesch-Kincaid grade level)
  - Required Elements (name greeting, company site, apply URL, CTA, email sign-off)
  - Spam & Formatting (spam words, excessive caps, exclamation marks, fake RE:)
  - Personalization Depth (verifies at least 1 real enriched attribute is referenced)
  - Entity Grounding Filter (scans for un-grounded numbers or stray proper nouns)
  - Rate Grounding Filter (verifies no rate figure is fabricated outside rate card lookup)
  - Placeholder Check (verifies zero unfilled placeholder tokens exist)

Send rule: All programmatic GATE checks pass.
"""

from __future__ import annotations

import re
from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List, Optional

from core import readability
from core.leads import Lead
from draft_generator import Draft
from logger import get_logger
from prompts import prompt_builder as pb

log = get_logger(__name__)

# Common cold-outreach spam-trigger words (deliverability signal)
SPAM_WORDS = {
    "100% free", "free money", "free trial", "free gift", "free offer", "claim your free",
    "guarantee", "guaranteed", "act now", "urgent", "risk-free",
    "limited time", "no obligation", "100%", "cash", "winner", "click here",
    "buy now", "cheap", "discount", "offer expires", "$$$",
}

CTA_VERBS = (
    "apply", "reply", "reach out", "get in touch", "connect", "submit",
    "explore", "chat", "call", "book", "let us know", "interested"
)


@dataclass
class Check:
    name: str
    passed: bool
    severity: str  # "gate" | "warn"
    detail: str
    value: Any = None


@dataclass
class Evaluation:
    channel: str
    lead_name: str
    checks: List[Check] = field(default_factory=list)
    programmatic_pass: bool = False
    send: bool = False
    flags: List[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        res = asdict(self)
        res["checks"] = [asdict(c) for c in self.checks]
        return res


def _spam_hit(term: str, body_lower: str) -> bool:
    if term.isalpha():
        return re.search(rf"\b{re.escape(term)}\b", body_lower) is not None
    return term in body_lower


def _caps_ratio(text: str) -> float:
    letters = [c for c in text if c.isalpha()]
    if not letters:
        return 0.0
    return sum(c.isupper() for c in letters) / len(letters)


def _lead_attribute_terms(draft: Draft) -> Dict[str, List[str]]:
    lead = draft.lead
    terms: Dict[str, List[str]] = {}
    langs = []
    for v in (lead.target_language, lead.source_language):
        if v:
            langs += [p.strip() for p in v.split(",") if p.strip()]
    langs += lead.secondary_languages
    if langs:
        terms["language"] = langs
    if lead.country:
        terms["country"] = [lead.country]
    if lead.services:
        terms["services"] = lead.services
    if lead.vendor_experience:
        terms["experience"] = [lead.vendor_experience]
    if lead.years_of_exp is not None:
        terms["experience"] = terms.get("experience", []) + [str(lead.years_of_exp)]
    return terms


def _heuristic_unsupported_entities(draft: Draft) -> List[str]:
    body = draft.body
    allowed = set()
    for v in draft.lead.grounding_facts().values():
        allowed |= {t.lower() for t in re.findall(r"[A-Za-z0-9]+", v)}
    allowed |= {"global3", "global", "resources", "team", "hi", "we", "i"}
    for v in pb.BRAND.values():
        allowed |= {t.lower() for t in re.findall(r"[A-Za-z0-9]+", str(v))}

    flags: List[str] = []
    body_no_urls = re.sub(r"https?://\S+", "", body).replace("Global3", "")
    for num in re.findall(r"\b\d{2,}\b", body_no_urls):
        if num.lower() not in allowed:
            flags.append(num)
    for phrase in re.findall(r"\b([A-Z][a-z]+(?: [A-Z][a-z]+)+)\b", body_no_urls):
        toks = [t.lower() for t in phrase.split()]
        if any(t not in allowed for t in toks):
            flags.append(phrase)
    name_toks = {t.lower() for t in (draft.lead.full_name or "").split()}
    return sorted({f for f in flags if f.lower() not in name_toks})


def evaluate(draft: Draft) -> Evaluation:
    """Run programmatic rule checks matching draft_poc/evaluator.py."""
    body = draft.body
    subject = draft.subject or ""
    lead = draft.lead
    checks: List[Check] = []
    flags: List[str] = []

    words = len(body.split())
    chars = len(body)

    # 1. Length ------------------------------------------------------------
    if draft.channel == "email":
        ok = 90 <= words <= 230
        ideal = 120 <= words <= 180
        checks.append(Check("length_words", ok, "gate",
            f"{words} words (band 90-230; ideal 120-180{'' if ideal else ' — outside ideal'})", words))
        if not ok:
            flags.append("LENGTH_OUT_OF_BOUNDS")
    else:
        ok = 60 <= chars <= 300
        checks.append(Check("length_chars", ok, "gate",
            f"{chars} chars (band 60-300; fits LinkedIn connection note cap)", chars))
        checks.append(Check("linkedin_note_cap", chars <= 300, "gate",
            f"{chars} chars ({'fits' if chars <= 300 else 'EXCEEDS'} 300-char cap)", chars))
        if not ok or chars > 300:
            flags.append("LINKEDIN_NOTE_CAP_EXCEEDED")

    # 2. Readability -------------------------------------------------------
    fre = readability.flesch_reading_ease(body)
    fk = readability.flesch_kincaid_grade(body)
    checks.append(Check("readability_flesch", fre >= 40, "warn",
        f"Flesch Reading Ease {fre:.0f} (>=40 good; ~60-70 = plain business English)", round(fre, 1)))
    checks.append(Check("readability_grade", fk <= 12, "warn",
        f"Flesch-Kincaid grade {fk:.1f} (<=12 target)", round(fk, 1)))

    # 3. Required elements -------------------------------------------------
    greets_name = lead.first_name.lower() in body.lower()[:60]
    has_apply = pb.BRAND["apply_url"] in body or "app.global3.io/apply" in body
    has_site = pb.BRAND["site"] in body
    has_cta = any(v in body.lower() for v in CTA_VERBS)
    if draft.channel == "email":
        has_signoff = "resources team" in body.lower()
        subj_ok = bool(subject) and len(subject.split()) <= 8
        req_ok = greets_name and has_apply and has_site and has_cta and has_signoff and subj_ok
        detail = (f"name✓{int(greets_name)} apply✓{int(has_apply)} site✓{int(has_site)} "
                  f"cta✓{int(has_cta)} signoff✓{int(has_signoff)} subject✓{int(subj_ok)}")
    else:
        req_ok = greets_name and has_apply and has_site and has_cta
        detail = (f"name✓{int(greets_name)} apply✓{int(has_apply)} "
                  f"site✓{int(has_site)} cta✓{int(has_cta)}")
    checks.append(Check("required_elements", req_ok, "gate", detail))
    if not req_ok:
        flags.append("MISSING_REQUIRED_ELEMENTS")

    # 4. Spam / formatting -------------------------------------------------
    spam_hits = [w for w in SPAM_WORDS if _spam_hit(w, body.lower())]
    caps = _caps_ratio(body)
    excls = body.count("!")
    fake_re = bool(re.match(r"\s*(re|fwd):", subject, re.I))
    spam_ok = len(spam_hits) == 0 and caps < 0.30 and excls <= 1 and not fake_re
    checks.append(Check("spam_formatting", spam_ok, "warn",
        f"spam_words={spam_hits or 'none'} caps={caps:.0%} '!'={excls} fake_re={fake_re}",
        len(spam_hits)))

    # 5. Personalization depth --------------------------------------------
    terms = _lead_attribute_terms(draft)
    hit_categories = [
        cat for cat, vals in terms.items()
        if any(val.lower() in body.lower() for val in vals if len(val) >= 3)
    ]
    depth = len(hit_categories)
    checks.append(Check("personalization_depth", depth >= 1, "gate",
        f"{depth} real attribute categorie(s) referenced: {hit_categories or 'none — only the name'}",
        depth))
    if depth < 1:
        flags.append("LOW_PERSONALIZATION_DEPTH")

    # 6. Entity grounding pre-filter --------------------------------------
    unsupported = _heuristic_unsupported_entities(draft)
    checks.append(Check("entity_grounding", not unsupported, "warn",
        f"possible unsupported specifics: {unsupported or 'none'}", len(unsupported)))

    # 7. Unfilled placeholders check ---------------------------------------
    placeholders = re.findall(r"(\[\s*[\w_]+\s*\]|\{\s*[\w_]+\s*\}|undefined|null|N/A)", body)
    no_placeholders = len(placeholders) == 0
    checks.append(Check("no_placeholders", no_placeholders, "gate",
        f"Unfilled placeholders: {placeholders or 'none'}"))
    if not no_placeholders:
        flags.append("UNFILLED_PLACEHOLDERS_FOUND")

    # 8. Rate grounding check ----------------------------------------------
    mentioned_rates = re.findall(r"\$\d+(?:\.\d{2})?|\d+\s*(?:USD|EUR|GBP|cents|per word)", body, re.I)
    if not draft.rate_match and mentioned_rates:
        rate_ok = False
        detail = f"Rate mentioned ({mentioned_rates}) but NO rate card match exists!"
        flags.append("FABRICATED_RATE_DETECTED")
    else:
        rate_ok = True
        detail = f"Rate match: {draft.rate_match['rate']} {draft.rate_match['currency']}" if draft.rate_match else "No rate mentioned"
    checks.append(Check("rate_grounding", rate_ok, "gate", detail))

    if draft.rate_flag == "NO_RATE_MATCH":
        flags.append("NO_RATE_MATCH")

    gates_pass = all(c.passed for c in checks if c.severity == "gate")

    return Evaluation(
        channel=draft.channel,
        lead_name=draft.lead.first_name,
        checks=checks,
        programmatic_pass=gates_pass,
        send=gates_pass,
        flags=flags,
    )

File: server/src/drafting/evaluator.ts
/** Evaluation layer — direct port of drafting_service/evaluator.py.
 *
 * Single-Stage Programmatic Evaluation:
 *   Rule-based checks (fast, transparent, fail-fast):
 *   - Length (words for email, characters for LinkedIn connection note cap)
 *   - Readability (Flesch Reading Ease & Flesch-Kincaid grade level)
 *   - Required Elements (name greeting, company site, apply URL, CTA, email sign-off)
 *   - Spam & Formatting (spam words, excessive caps, exclamation marks, fake RE:)
 *   - Personalization Depth (verifies at least 1 real enriched attribute is referenced)
 *   - Entity Grounding Filter (scans for un-grounded numbers or stray proper nouns)
 *   - Rate Grounding Filter (verifies no rate figure is fabricated outside rate card lookup)
 *   - Placeholder Check (verifies zero unfilled placeholder tokens exist)
 *
 * Send rule: All programmatic GATE checks pass. */

import { fleschKincaidGrade, fleschReadingEase } from "./readability";
import { specificityTarget, type Draft } from "./draftGenerator";
import { BRAND, LINKEDIN_NOTE_CHAR_CAP } from "./promptBuilder";
import { citedNamedSpecifics } from "./leads";

// Common cold-outreach spam-trigger words (deliverability signal)
const SPAM_WORDS = [
  "100% free", "free money", "free trial", "free gift", "free offer", "claim your free",
  "guarantee", "guaranteed", "act now", "urgent", "risk-free",
  "limited time", "no obligation", "100%", "cash", "winner", "click here",
  "buy now", "cheap", "discount", "offer expires", "$$$",
];

const CTA_VERBS = [
  "apply", "reply", "reach out", "get in touch", "connect", "submit",
  "explore", "chat", "call", "book", "let us know", "interested",
];

export interface Check {
  name: string;
  passed: boolean;
  severity: "gate" | "warn";
  detail: string;
  value?: any;
}

export interface Evaluation {
  channel: string;
  lead_name: string;
  checks: Check[];
  programmatic_pass: boolean;
  send: boolean;
  flags: string[];
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function spamHit(term: string, bodyLower: string): boolean {
  if (/^[A-Za-z]+$/.test(term)) {
    return new RegExp(`\\b${escapeRegExp(term)}\\b`).test(bodyLower);
  }
  return bodyLower.includes(term);
}

function capsRatio(text: string): number {
  const letters = Array.from(text).filter((c) => /\p{L}/u.test(c));
  if (!letters.length) return 0.0;
  const upperCount = letters.filter((c) => /\p{Lu}/u.test(c)).length;
  return upperCount / letters.length;
}

function leadAttributeTerms(draft: Draft): Record<string, string[]> {
  const lead = draft.lead;
  const terms: Record<string, string[]> = {};
  let langs: string[] = [];
  for (const v of [lead.targetLanguage, lead.sourceLanguage]) {
    if (v) langs = langs.concat(v.split(",").map((p) => p.trim()).filter(Boolean));
  }
  langs = langs.concat(lead.secondaryLanguages);
  if (langs.length) terms.language = langs;
  if (lead.country) terms.country = [lead.country];
  if (lead.services.length) terms.services = lead.services;
  if (lead.vendorExperience) terms.experience = [lead.vendorExperience];
  if (lead.yearsOfExp !== null) terms.experience = (terms.experience || []).concat([String(lead.yearsOfExp)]);
  return terms;
}

function heuristicUnsupportedEntities(draft: Draft): string[] {
  const body = draft.body;
  const allowed = new Set<string>();
  for (const v of Object.values(draft.lead.groundingFacts())) {
    for (const t of v.match(/[A-Za-z0-9]+/g) || []) allowed.add(t.toLowerCase());
  }
  // The raw enrichment payloads are handed to the model as a sanctioned place
  // to find one more distinctive detail (see promptBuilder's ADDITIONAL RAW
  // PROFILE DATA block), so anything named in them is grounded -- it just
  // isn't in the curated facts. Without this, the check contradicts the
  // prompt and flags the model for doing exactly what it was told: confirmed
  // live 2026-09-07, "Tata Mutual Fund" was flagged as an unsupported
  // specific for a lead whose own Parallel payload named it.
  for (const payload of [draft.lead.parallelFullData, draft.lead.rawScrapeData]) {
    if (!payload) continue;
    for (const t of JSON.stringify(payload).match(/[A-Za-z0-9]+/g) || []) allowed.add(t.toLowerCase());
  }
  for (const w of ["global3", "global", "resources", "team", "hi", "we", "i"]) allowed.add(w);
  for (const v of Object.values(BRAND)) {
    for (const t of String(v).match(/[A-Za-z0-9]+/g) || []) allowed.add(t.toLowerCase());
  }

  const flags: string[] = [];
  const bodyNoUrls = body.replace(/https?:\/\/\S+/g, "").split("Global3").join("");
  for (const num of bodyNoUrls.match(/\b\d{2,}\b/g) || []) {
    if (!allowed.has(num.toLowerCase())) flags.push(num);
  }
  // Multi-word names are checked as a WHOLE PHRASE against everything known
  // about the lead, NOT token by token. Token-level matching let a fabricated
  // pairing through whenever both halves happened to appear somewhere
  // unrelated -- "Sony Pictures" passed for a lead whose payload mentioned
  // "Sony" and, separately, "Pictures" -- and pouring both raw payloads into
  // the token pool widened exactly that hole.
  //
  // The skip list here is deliberately ONLY our own brand wording and the
  // lead's own name: those are the phrases that legitimately appear in a
  // draft without being grounded in profile data. Skipping anything whose
  // tokens are individually "allowed" would reintroduce the same bug, since
  // that set contains every token of both payloads.
  const structural = new Set<string>();
  for (const w of ["global3", "global", "resources", "team", "hi", "we", "i", "best", "regards"]) structural.add(w);
  for (const v of Object.values(BRAND)) {
    for (const t of String(v).match(/[A-Za-z0-9]+/g) || []) structural.add(t.toLowerCase());
  }
  for (const t of `${draft.lead.fullName || ""} ${draft.lead.firstName || ""}`.match(/[A-Za-z0-9]+/g) || []) {
    structural.add(t.toLowerCase());
  }
  const known = draft.lead.allKnownText();
  for (const match of bodyNoUrls.matchAll(/\b([A-Z][a-z]+(?: [A-Z][a-z]+)+)\b/g)) {
    const phrase = match[1];
    const toks = phrase.split(" ").map((t) => t.toLowerCase());
    if (toks.every((t) => structural.has(t))) continue;
    if (!known.includes(phrase.toLowerCase())) flags.push(phrase);
  }
  const nameToks = new Set(
    (draft.lead.fullName || "").split(/\s+/).filter(Boolean).map((t) => t.toLowerCase())
  );
  const uniqueFlags = new Set(flags.filter((f) => !nameToks.has(f.toLowerCase())));
  return Array.from(uniqueFlags).sort();
}

/** Run programmatic rule checks matching drafting_service/evaluator.py. */
export function evaluate(draft: Draft): Evaluation {
  const body = draft.body;
  const subject = draft.subject || "";
  const lead = draft.lead;
  const checks: Check[] = [];
  const flags: string[] = [];

  const words = body.split(/\s+/).filter(Boolean).length;
  const chars = body.length;

  // 1. Length ----------------------------------------------------------
  if (draft.channel === "email") {
    const ok = words >= 90 && words <= 230;
    const ideal = words >= 120 && words <= 180;
    checks.push({
      name: "length_words",
      passed: ok,
      severity: "gate",
      detail: `${words} words (band 90-230; ideal 120-180${ideal ? "" : " — outside ideal"})`,
      value: words,
    });
    if (!ok) flags.push("LENGTH_OUT_OF_BOUNDS");
  } else {
    const ok = chars >= 60 && chars <= LINKEDIN_NOTE_CHAR_CAP;
    checks.push({
      name: "length_chars",
      passed: ok,
      severity: "gate",
      detail: `${chars} chars (band 60-${LINKEDIN_NOTE_CHAR_CAP}; fits LinkedIn connection-note cap)`,
      value: chars,
    });
    checks.push({
      name: "linkedin_note_cap",
      passed: chars <= LINKEDIN_NOTE_CHAR_CAP,
      severity: "gate",
      detail: `${chars} chars (${chars <= LINKEDIN_NOTE_CHAR_CAP ? "fits" : "EXCEEDS"} ${LINKEDIN_NOTE_CHAR_CAP}-char cap)`,
      value: chars,
    });
    if (!ok || chars > LINKEDIN_NOTE_CHAR_CAP) flags.push("LINKEDIN_NOTE_CAP_EXCEEDED");
  }

  // 2. Readability -------------------------------------------------------
  const fre = fleschReadingEase(body);
  const fk = fleschKincaidGrade(body);
  checks.push({
    name: "readability_flesch",
    passed: fre >= 40,
    severity: "warn",
    detail: `Flesch Reading Ease ${fre.toFixed(0)} (>=40 good; ~60-70 = plain business English)`,
    value: Math.round(fre * 10) / 10,
  });
  checks.push({
    name: "readability_grade",
    passed: fk <= 12,
    severity: "warn",
    detail: `Flesch-Kincaid grade ${fk.toFixed(1)} (<=12 target)`,
    value: Math.round(fk * 10) / 10,
  });

  // 3. Required elements -------------------------------------------------
  const greetsName = body.toLowerCase().slice(0, 60).includes(lead.firstName.toLowerCase());
  const hasApply = body.includes(BRAND.apply_url) || body.includes("app.global3.io/apply");
  const hasSite = body.includes(BRAND.site);
  const hasCta = CTA_VERBS.some((v) => body.toLowerCase().includes(v));
  let reqOk: boolean;
  let detail: string;
  if (draft.channel === "email") {
    const hasSignoff = body.toLowerCase().includes("resources team");
    const subjOk = !!subject && subject.split(/\s+/).filter(Boolean).length <= 8;
    reqOk = greetsName && hasApply && hasSite && hasCta && hasSignoff && subjOk;
    detail = `name✓${Number(greetsName)} apply✓${Number(hasApply)} site✓${Number(hasSite)} cta✓${Number(hasCta)} signoff✓${Number(hasSignoff)} subject✓${Number(subjOk)}`;
  } else {
    reqOk = greetsName && hasApply && hasSite && hasCta;
    detail = `name✓${Number(greetsName)} apply✓${Number(hasApply)} site✓${Number(hasSite)} cta✓${Number(hasCta)}`;
  }
  checks.push({ name: "required_elements", passed: reqOk, severity: "gate", detail });
  if (!reqOk) flags.push("MISSING_REQUIRED_ELEMENTS");

  // 4. Spam / formatting -------------------------------------------------
  const bodyLower = body.toLowerCase();
  const spamHits = SPAM_WORDS.filter((w) => spamHit(w, bodyLower));
  const caps = capsRatio(body);
  const excls = (body.match(/!/g) || []).length;
  const fakeRe = /^\s*(re|fwd):/i.test(subject);
  const spamOk = spamHits.length === 0 && caps < 0.3 && excls <= 1 && !fakeRe;
  checks.push({
    name: "spam_formatting",
    passed: spamOk,
    severity: "warn",
    detail: `spam_words=${spamHits.length ? JSON.stringify(spamHits) : "none"} caps=${(caps * 100).toFixed(0)}% '!'=${excls} fake_re=${fakeRe}`,
    value: spamHits.length,
  });

  // 5. Personalization depth --------------------------------------------
  const terms = leadAttributeTerms(draft);
  const hitCategories = Object.entries(terms)
    .filter(([, vals]) => vals.some((val) => val.length >= 3 && bodyLower.includes(val.toLowerCase())))
    .map(([cat]) => cat);
  const depth = hitCategories.length;
  // A cited named detail satisfies this too. This check matches the body
  // against the CANONICAL columns (language/service/country/years), and those
  // are routinely empty or plain wrong on real rows -- confirmed live
  // 2026-09-07: a lead whose record said `targetLanguage: "English",
  // services: []` while his actual profile headline read "Voice & Dubbing
  // Artist Punjabi Hindi" got a correctly personalized draft ("your Hindi &
  // Punjabi dubbing work stood out") and then failed this gate, because
  // "Hindi" matches nothing in a record that claims English. Naming a
  // grounded detail from the profile is strictly STRONGER evidence of
  // personalization than echoing a canonical attribute, so it counts here as
  // well; a draft with neither is still gated, which is the mail-merge case
  // this exists to catch.
  const specificsForDepth = citedNamedSpecifics(lead, body);
  const depthOk = depth >= 1 || specificsForDepth.length >= 1;
  checks.push({
    name: "personalization_depth",
    passed: depthOk,
    severity: "gate",
    detail:
      `${depth} canonical attribute categorie(s) referenced: ${depth ? JSON.stringify(hitCategories) : "none"}` +
      `; ${specificsForDepth.length} named profile detail(s) cited`,
    value: depth,
  });
  if (!depthOk) flags.push("LOW_PERSONALIZATION_DEPTH");

  // 5b. Named-specificity depth ------------------------------------------
  // personalization_depth above only proves a broad attribute (a language, a
  // service category, a country) was mentioned -- "your subtitling experience
  // in Spain" passes it while reading like a mail merge. This check is the
  // one that asks whether the draft names things only THIS person's profile
  // could have supplied: a tool, a credential, an employer, a production.
  // Bar is min(2, available) for email and min(1, available) for LinkedIn,
  // whose ~200-char note has room for one -- so a genuinely thin profile is
  // never held to a standard its own data can't meet. Shares
  // Lead.specificFactCandidates() with draftGenerator's regenerate-once
  // guard so the generator and the evaluator can't disagree on the bar.
  const strongFacts = lead.strongFactCandidates();
  const citedFacts = citedNamedSpecifics(lead, body);
  // Email asks for the full bar; LinkedIn's ~200-char note has room for one.
  const specTarget =
    draft.channel === "email" ? specificityTarget(strongFacts) : Math.min(1, strongFacts.length);
  const specOk = citedFacts.length >= specTarget;
  checks.push({
    name: "named_specificity",
    passed: specOk,
    severity: "gate",
    detail:
      strongFacts.length === 0
        ? "no named org/tool/credential on this lead's profile — bar waived"
        : `${citedFacts.length}/${specTarget} named detail(s) cited${citedFacts.length ? ": " + JSON.stringify(citedFacts.slice(0, 4)) : ""} (profile offers ${strongFacts.length} strong)`,
    value: citedFacts.length,
  });
  if (!specOk) flags.push("LOW_NAMED_SPECIFICITY");

  // 6. Entity grounding pre-filter ---------------------------------------
  const unsupported = heuristicUnsupportedEntities(draft);
  checks.push({
    name: "entity_grounding",
    passed: unsupported.length === 0,
    severity: "warn",
    detail: `possible unsupported specifics: ${unsupported.length ? JSON.stringify(unsupported) : "none"}`,
    value: unsupported.length,
  });

  // 7. Unfilled placeholders check ---------------------------------------
  // Also matches MULTI-WORD bracket slots: the email exemplar now uses
  // "[NAMED DETAIL #2: named tools, a credential, ...]" placeholders, and a
  // single-token-only pattern would let a leaked one through as send-ready.
  const placeholders = body.match(/(\[[^\]\n]{1,120}\]|\{\s*[\w_]+\s*\}|undefined|null|N\/A)/g) || [];
  const noPlaceholders = placeholders.length === 0;
  checks.push({
    name: "no_placeholders",
    passed: noPlaceholders,
    severity: "gate",
    detail: `Unfilled placeholders: ${placeholders.length ? JSON.stringify(placeholders) : "none"}`,
  });
  if (!noPlaceholders) flags.push("UNFILLED_PLACEHOLDERS_FOUND");

  // 8. Rate grounding check ----------------------------------------------
  const mentionedRates = body.match(/\$\d+(?:\.\d{2})?|\d+\s*(?:USD|EUR|GBP|cents|per word)/gi) || [];
  let rateOk: boolean;
  let rateDetail: string;
  if (!draft.rate_match && mentionedRates.length) {
    rateOk = false;
    rateDetail = `Rate mentioned (${JSON.stringify(mentionedRates)}) but NO rate card match exists!`;
    flags.push("FABRICATED_RATE_DETECTED");
  } else {
    rateOk = true;
    rateDetail = draft.rate_match ? `Rate match: ${draft.rate_match.rate} ${draft.rate_match.currency}` : "No rate mentioned";
  }
  checks.push({ name: "rate_grounding", passed: rateOk, severity: "gate", detail: rateDetail });

  if (draft.rate_flag === "NO_RATE_MATCH") flags.push("NO_RATE_MATCH");

  const gatesPass = checks.filter((c) => c.severity === "gate").every((c) => c.passed);

  return {
    channel: draft.channel,
    lead_name: draft.lead.firstName,
    checks,
    programmatic_pass: gatesPass,
    send: gatesPass,
    flags,
  };
}
